### PR TITLE
raster: Restrict GDAL-linked raster reads to the needed column range

### PR DESCRIPTION
**Pull request by ninsbl**
_read_data_gdal()_ - used when reading a linked-raster maps (**r.external**) - always reads the full native row width from the GDAL dataset, regardless of the current region.
For a raster map that is much wider than the region - e.g. a country-wide mosaic VRT built from many source tiles, this can have significant impact on the processing time. Because every row read touches every source tile across the full width, even when the region only overlaps one or two of them. The consequence is significantly degraded performance for such cases, esp. if several maps are read simultaneously e.g. in a mapcalculator expression and if data is stored on file systems with higher latency (e.g. NAS with CIFS mount).

This PR restricts the GDAL read to the native column range actually needed. It is computed from the column mapping GRASS already builds once per file open. The behavior for hflip'ed datasets is unchanged and falls back to the original full-width read.

**r.in.gdal** and **r.external** are unaffected by the changes. OpenMP parallelized tools are not affected by the change either. I do have a regression-test for that in my local working copy and can push it to the PR if it seems useful. And I am happy to remove the r.in.gdal test if that is not needed (I just wanted to make sure all possible caveats are covered).

The PR includes a pytest covering cases where the computaional region is fully inside, partially outside, and fully outside the linked file's extent. THere is also one test to verify **r.in.gdal** is unaffected.

Given the significant performance degradation in some cases I take the liberty to classify this as a blocker. May be also a reason to backport...

Generated with the help of AI with lots of manual debugging and verifications...

### CODE REVIEW

File: lib/raster/get_row.c [hunk 1]
```diff
@@ -207,26 +207,36 @@ static void read_data_gdal(int fd, int row, unsigned char *data_buf,
     struct fileinfo *fcb = &R__.fileinfo[fd];
     unsigned char *buf;
     CPLErr err;
+    /* Restrict the read to the native columns actually needed by the
+     * region (computed once in Rast__create_window_mapping()) instead
+     * of always reading the full native row width. hflip'ed maps keep the
+     * original full-width behavior, to avoid also having to mirror
+     * the needed column range. */
```
Comment (echoix) [inline lib/raster/get_row.c]: Since using AI, ask it to be more concise here, imagine if all our code had that much explanations.

File: lib/raster/tests/lib_raster_gdal_link_window_test.py
Comment (echoix) [inline lib/raster/tests/lib_raster_gdal_link_window_test.py:1]: Is this the regression test you were talking about, or you have more to add?

**Comment by echoix** [comment]
I don't understand the C code changes enough to approve or reject

File: lib/raster/tests/lib_raster_gdal_link_window_test.py
Comment (ninsbl) [reply lib/raster/tests/lib_raster_gdal_link_window_test.py:1]: I have more to add, if considered useful / needed:

1. parallel vs. serial reads with r.univar with nprocs=1 and nprocs>1
2. tests for resolution mismatch (computational region region vs raster file)
3. wrapping around meridian in WGS84 CRS

All those three tests pass and I am happy to push them to the PR if anyone thiks they should be added. Also, I am happy to check other possible edge cases if anyone can think of other cases to be covered.

This is quite central code, so regressions here would be particularly nasty...

**Comment by ninsbl** [comment]
> I don't understand the C code changes enough to approve or reject

Thanks for looking into it anyway!

**Comment by ninsbl** [comment]
If someone with knowledge of this area of the code could review this PR, I would be really glad. I am aware of time constraints of course and I would not ask if not this would be a major issue (for me). It grinds one of my use-cases for GRASS practically to a halt, because stuff that should take less thqan half a minute can take more than two hours...

So, this PR would increase the performance of GRASS for my case up to more than x250-300 (depending on the input dataset and file system of course)...

I assume it will also significantly improve usage of GRASS as processing backend in QGIS in many cases...

**Review by echoix — DISMISSED** [review]
Well, it doesn’t look wrong, and you’ll be the first to change it back if it doesn’t work, so here you go.

I was hoping on someone that would have known if something was missing or could be done in a better way if it existed.

File: lib/raster/get_row.c (hunk 1 above)
Comment (ninsbl) [reply lib/raster/get_row.c]: Done.

File: lib/raster/tests/lib_raster_gdal_link_window_test.py
Comment (ninsbl) [reply lib/raster/tests/lib_raster_gdal_link_window_test.py:1]: With [d438148](https://github.com/OSGeo/grass/pull/7876/commits/d43814817449c0cebe241e4c7cc19cabbeed7ff8), I added the tests I mentioned. That may be covered other places, but I wanted to make sure that things work as expected with linked data also after the change. Tests take about 4 seconds on my laptop...

**Comment by ninsbl** [comment]
> Well, it doesn’t look wrong, and you’ll be the first to change it back if it doesn’t work, so here you go.
> 
> I was hoping on someone that would have known if something was missing or could be done in a better way if it existed.

Thanks! I addressed your comments and added the mentioned tests...

**Comment by nilason** [comment]
The code changes looks reasonable to me, but I’m not totally comfortable with this part of the code, let us give @metzm a chance to chime in.

**Comment by echoix** [comment]
I ran an AI review over this PR (Opus 5 though) and had it verify the asserted values by compiling `lib/arraystats/{basic,class}.c` and `lib/rowio/*.c` standalone against stubbed libgis. Good news first: every expected value in the three test files is correct, and the ctypes plumbing is sound.

Two things are probably worth acting on:

- **A real stack-buffer-overflow read in `AS_class_frequencies`, confirmed under ASan** (`class.c:450`), reachable from the library's own output — the new frequencies test just misses it.
- **One test asserts a guarantee `rowio` does not actually provide**, so it passes for the wrong reason and the branch it claims to cover never runs.

Details, plus smaller coverage gaps:

<details>
<summary>Full findings</summary>

### Blocking

**1. `lib/rowio/tests/lib_rowio_ctypes_test.py:170` — `test_get_returns_none_when_getrow_fails` asserts a guarantee the library does not provide.**

The test passes only because no successful `get` precedes it. `Rowio_get`'s cleanup guard is

```c
if (cur == R->cur)
    R->cur = -1;
```

which compares a *slot index* (`cur`) against a *row number* (`R->cur`), so it usually doesn't fire. Confirmed in a harness: with `nrows=1`, `get(3)` followed by a failing `get(7)` leaves `R->cur == 3`, and a later `get(3)` returns the clobbered buffer (`"hhhh"` instead of `"dddd"`) without calling `getrow` at all. So the docstring's claim that the cache is "not left in a half-cached state" is false, and the branch the commit message says this test covers never executes. Either rewrite the test to document the actual (buggy) behaviour, or fix the guard in `rowio.c` and extend the test to do a successful get first.

**2. `lib/arraystats/tests/lib_arraystats_classify_ctypes_test.py:79` — the frequencies test can't reach the out-of-bounds read.**

It passes only interior breaks, so the unbounded `while (data[i] <= classbreaks[j])` loop in `AS_class_frequencies` is never exercised. It is reachable from the library's own output: `AS_class_quant({1,2,3,4}, 4, 3, ...)` yields breaks `{2,3,4}`, and feeding those back in gives a stack-buffer-overflow read at `class.c:450` under ASan.

**3. `lib/arraystats/tests/lib_arraystats_classify_ctypes_test.py:88` — `test_class_discont_on_a_uniform_range` is toolchain-dependent.**

Perfectly uniform data is the one input where the residual `d` is mathematically 0, so the `d <= dmax` comparison is decided by ~1e-17 of rounding noise. Measured: gcc `-O0`/`-O2`/`-O3 -march=native` and clang all give `0.125` / `[2.5, 5.5]`; gcc with `-ffast-math` gives `0.316` / `[2.5, 10.0]`. The same pin exists in the `CLASS_DISCONT` row of the dispatcher test. Non-uniform data would keep the branch outcome from being decided by the compiler.

### Coverage gaps

- **`classify_ctypes_test.py:88`** — no degenerate-input discont case. All-equal values, or `count == 1`, give `rangemax == 0` → 0/0 → NaN breaks plus the sentinel `chi2 = 1000`, and `AS_class_apply_algorithm`'s `finfo == 0` check does not catch it (harness-verified).
- **`classify_ctypes_test.py:69`** — `AS_class_equiprob`'s class-reduction branch (the warning, the partial zeroing, the `*nbreaks` rewrite) is untested. `[1]*9 + [100]` with `nbreaks=9` triggers it.
- **`classify_ctypes_test.py:131`** — `byref(c_int(nbreaks))` creates a throwaway, so the `CLASS_EQUIPROB` case cannot observe the `*nbreaks` write-back, which is the entire reason that parameter is a pointer. Binding the `c_int` to a local would let it be asserted.
- **`basic_stats_ctypes_test.py:26`** — only 4 of the 10 `GASTATS` fields are asserted, and all inputs are positive, so `sum`/`sumabs` and `mean`/`meanabs` are indistinguishable; swapping them in `basic.c` would keep the suite green. A negative-value case plus the remaining fields would fix both.
- **`basic_stats_ctypes_test.py:58`** — `c`/`rc` is declared but never asserted, so a regression that spuriously sets `*c` for a non-vertical line passes. `AS_class_discont` branches on exactly that value.
- **`lib_rowio_ctypes_test.py:56`** — missing the negative-row guards (`Rowio_get(-1)` → NULL, `Rowio_put(..., -1)` → 0) and the documented `putrow == NULL` configuration, which `pageout()` would otherwise call unconditionally on a dirty row.

### Note (not blocking)

**`lib_rowio_ctypes_test.py:93`** — these LRU assertions are the first code to depend on `rcb[].age`, which `Rowio_setup` never initializes (it `G_malloc`s and sets only `.row`), and `my_select` bumps the age of unused slots on every get. It's benign at `nrows=2` because the free-slot `break` wins first, but any later case with `nrows >= 3` that evicts while slots are still unused would be nondeterministic. Worth a comment so the next person doesn't extend it.

### Verified as correct

min/max/mean/stdev, interval, quant, stdev, equiprob, the frequency counts, and the discont chi2/breaks all reproduce. The `CFUNCTYPE` signatures match ctypesgen's generated argtypes, both libraries are in the interface-generator module list, the filenames match the `*/tests/*_test.py` discovery pattern in `pyproject.toml`, and the style follows the `lib/datetime/tests` precedent.

</details>

**Comment by ninsbl** [comment]
@echoix just to make sure: Your AI review referenced only tests outside of this PR. Are the findings related to the changes here or to the (relativly new) ctypes based tests?

**Comment by echoix** [comment]
> @echoix just to make sure: Your AI review referenced only tests outside of this PR. Are the findings related to the changes here or to the (relativly new) ctypes based tests?

Oh, did I mistype the PR number ? I didn't caught on since it was the same topics and touching bounds. Let me retry something. Or you could on your side too.

**Comment by echoix** [comment]
> > @echoix just to make sure: Your AI review referenced only tests outside of this PR. Are the findings related to the changes here or to the (relativly new) ctypes based tests?
> 
> Oh, did I mistype the PR number ? I didn't caught on since it was the same topics and touching bounds. Let me retry something. Or you could on your side too. 

Indeed, it's for #7879

**Comment by echoix** [comment]
### AI-assisted review — Claude Code (Opus 5)

**No correctness bugs found.** Reviewed `775c8eb` against `origin/main` (4 files, +267/−11): the narrowed `GDALRasterIO` window is index-consistent with every consumer of `fcb->data`, and the cached column range cannot go stale because `Rast_set_input_window()` already fatal-errors while any map is open for read. Two things worth a look: the `-1` sentinel for `gdal_min_col`/`gdal_max_col` is not what an fd actually defaults to (`new_fileinfo()` zeroes the struct, so a skipped mapping reads as the valid range "column 0 only"), and nothing in the new tests exercises an `hflip`/`vflip` link — the one branch the change deliberately special-cases.

<details>
<summary>Full review</summary>

#### What was verified

The change is only safe if a handful of invariants hold, so each was checked against the source rather than by eye:

- **Buffer layout stays consistent.** `read_data_gdal()` writes native column `c` at `data_buf + c * cur_nbytes` (absolute, not relative to `col_off`), which is exactly where `gdal_values_int/float/double()` read it via `col_map[i] - 1` (`get_row.c:365,412,438`). Every consumer of `fcb->data`, `get_null_value_row_gdal()` included, goes through `col_map`, so untouched columns are never read.
- **Bounds.** `col_map` entries are clamped to `[0, cellhd.cols]` in `Rast__create_window_mapping()`, so `gdal_min_col`/`gdal_max_col` land in `[0, cols-1]`; the write range fits the `G_calloc(cellhd.cols, MAP_NBYTES)` buffer at `open.c:337`. `COLUMN_MAPPING` is `typedef int` (`R.h:16`), so the `-1` sentinel and the `>= 0` tests are signed-safe.
- **The cached range cannot go stale.** `Rast__create_window_mapping()` is called from `open.c:327`, after `fcb->gdal` and `fcb->cellhd` are set and while `open_mode == -1`, so the write-mode early return does not fire. `Rast_set_window()` / `Rast_set_input_window()` fatal-error in `update_window_mappings()` if any map is open for read, so the cached min/max has exactly the same lifetime as `col_map` itself.
- **Fallbacks.** No overlap gives `gdal_min_col == -1` and a full-width read. `hflip` is excluded in both places, and with `ncols == cellhd.cols` the rewritten flip loop is byte-for-byte the old one. VRT maps skip the mapping, but `get_map_row_nomask()` returns early on `fcb->vrt` before `read_data()`, and `get_row.c:595` is the only `read_data()` call site.
- **`cur_nbytes` is current** at the new pointer arithmetic: `*nbytes = fcb->nbytes` runs first and `nbytes` aliases `&fcb->cur_nbytes` at the sole call site. Fragile aliasing, but the pre-existing `hflip` `G_malloc` already depended on it.
- **The parallel tests are meaningful, not racy.** `r.mapcalc` opens a separate fd per thread (`raster/r.mapcalc/map.c:581`, keyed on `thread_num`) and `Rast_get_gdal_link()` does its own `GDALOpen` per fd (`gdal.c:137`), so no `GDALRasterBandH` is shared across threads.
- **Test math.** All six region cases were hand-checked against the C mapping (`C2` accumulation, the negative-rounding `x--`, and the lat/lon wraparound second pass). The antimeridian case resolves to native columns `350..359, 0..9`, matching the helper.

#### Findings

- **`lib/raster/get_row.c:213` / `lib/raster/open.c:34` — the `-1` sentinel is not the actual default.** `new_fileinfo()` `memset`s the struct to zero, so any fd where `Rast__create_window_mapping()` does not run leaves `gdal_min_col == gdal_max_col == 0`, which reads as the *valid* range "native column 0 only" rather than "unknown". Today the only such fds are VRT and write-mode maps, neither of which reaches `read_data_gdal()`, so there is no failure now. But a future read path that skips the mapping would silently read one column per row instead of the full width, producing plausible-looking wrong data rather than crashing. Initializing both to `-1` in `new_fileinfo()`, or at the top of `Rast__create_window_mapping()` before its early return, would make the sentinel self-defending.
- **`lib/raster/window_map.c:110` — no coverage for the branch the change special-cases.** `hflip` is excluded in two places and `vflip` remaps the row, yet no test creates a flipped link. A `vflip` link read under a restricted region is cheap insurance and covers the one interaction (`col_off` offsetting plus row inversion) that is currently untested.
- **`lib/raster/tests/lib_raster_gdal_link_window_test.py:180` — `test_r_in_gdal_ignores_region` does not touch the changed code.** `r.in.gdal` imports into native GRASS format and never goes through `read_data_gdal()`, so this asserts library behavior unrelated to the PR in a file whose docstring scopes it to windowed GDAL-linked reads.
- **`lib/raster/tests/lib_raster_gdal_link_window_test.py:220` — double space in the docstring** ("r.univar with&nbsp;&nbsp;linked data"), which `AGENTS.md` and `doc/development/style_guide.md` disallow.

#### Notes, not findings

- `R.h` reflows the comment alignment on three untouched lines (`reclass_flag`, `row_ptr`, `col_map`) — clang-format fallout from the longer new field names, expected.
- The new module-level `session` fixture shadows the one in `lib/raster/tests/conftest.py`. Legal pytest and module-local, but worth knowing it is intentional.
- Unrelated pre-existing issue noticed while checking the stride assumption: `raster/r.external/link.c:34` maps `GDT_Int8` to `cellhd->format = 1` (2 bytes/cell) while GDAL packs Int8 at 1 byte, so Int8 links are already misread. Not this PR's to fix, and the PR neither introduces nor worsens it.

#### Method and limits

Static review only — GRASS was not built and the new test file was not executed. The review was run by a human-directed AI agent; all findings above were traced to the cited source lines by hand.

</details>

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ninsbl** [comment]
OK, now I tried to address the review by adding tests, and initializing gdal_min_col and gdal_max_col at file opening.
Finally, I also included flipped maps in the read optimization (after I first added tests for flipped maps).

**Review by echoix — APPROVED** [review]
